aquadoggo/src/db/models/operation.rs

```diff
@@ -25,6 +25,17 @@ pub struct OperationRow {
     /// The previous operations of this operation concatenated into string format with `_`
     /// separator.
     pub previous: Option<String>,
+
+    /// Index for the position of this operation once topological sorting of the operation graph
+    /// has been performed.
+    ///
+    /// This is an option as when an operation which is not appended to the tip of a
+    /// document graph is handled then a `reduce` materialization task will be issued and all
+    /// indexes for operations in the effected document re-calculated.
+    ///
+    /// If this value is None we can assume the operation has not been processed yet and we are
+    /// waiting for the reduce task to complete.
+    pub sorted_index: Option<i32>,
```

Comment:
If this is `None` and we restart our node, would we still be able to materialize? I assume not, as the order (defined by the remote node we've received the data from) is lost?

Reply:
If that is the case we maybe directly set the ordering on arrival but introduce a flag for if it has been materialized

Reply:
Or we store the arrival order/ timestamp in our tasks table and reissue the tasks in the right order when the node restarts, that's probably the easiest

Reply:
> If this is None and we restart our node, would we still be able to materialize? I assume not, as the arrival order (defined by the remote node we've received the data from) is lost?

Actually, that's not the case. We don't use the order that the operations are supplied in for anything, we simply call `publish` in order and then as the subsequent `reduce` tasks complete, then `sorted_index` gets set for the new operations.

So we are actually leaning on our nice task re-run logic which already exists 👍

Reply:
> Actually, that's not the case. We don't use the order that the operations are supplied in for anything except to call publish in the correct order. As the subsequent reduce tasks complete, sorted_index gets set for the new operations.

I mean the case when `reduce` has not been completed ..

Not sure if `tasks` currently returns tasks in the order they arrived on restart, that needs checking maybe.

Reply:
Ah! Sorry, I see now. Somehow thought that publish comes later during replication 🙈

Reply:
> I mean the case when reduce has not been completed.

Hmm, yeh, I was meaning that too. Maybe I'm misunderstanding something. I thought:

- operations arrive, they're stored, `reduce` tasks issued (possibly de-duplication happens)
=====CRASH=====
- incomplete `reduce` tasks re-issued
- the order they're processed in doesn't matter, in any case all available operations will be included in the materialised document, and operations will have their `sorted_index` updated